lib/recursive-methods.rb

```diff
@@ -1,49 +1,108 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(n)
+# source: video example in class
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0
+        raise ArgumentError
+    elsif n == 1 || n == 0
+        return 1 
+    else
+        return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s 
+    else
+        return s[-1] + reverse(s[1...-1]) + s[0]
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: 0(n)
+# Space complexity: 0(1)
+def reverse_inplace(s, low = 0, high = s.length - 1) 
+    if low >= high
+        return s
+    else
+        temp = s[high]
+        s[high] = s[low]
+        s[low] = temp
+    end
+    
+    return reverse_inplace(s, low +=1, high -= 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    if n == 0
+        return 0
+    else
+        return 2 + bunny(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
+# Time complexity: 0n
+# Space complexity: 0(n^2)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+    if s == ""
+        return true
+    elsif s[0] == "(" && s[-1] == ")" && nested(s[1...-1])
+         return true
+    else 
+        return false
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n^2)
+# Space complexity: 0(n^2) <-- becuase I'm making a new array
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+    if array.length == 0
+        return false
+    elsif value == array[0]
+        return true
+    else 
+        return search(array[1..-1], value)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
+# Time complexity: 0(n^2)
+# Space complexity: 0(n^2) <-- becuase I'm making a new array
 def is_palindrome(s)
```

Comment:
👍 

Could you make a more efficient version with an index parameter like the reverse in place method.